File: (1) borrow_checker_basics/main.cpp
#include <iostream>
#include <vector>

struct Foo {
public:
    Foo(std::vector<unsigned>&& buffer) : buffer(std::move(buffer)) {}

private:
    std::vector<unsigned> buffer;
};

void add_zero(std::vector<unsigned> &input) { 
    input.push_back(1); 
}

std::vector<unsigned>&& take_and_return(std::vector<unsigned>&& input) {
    add_zero(input);
    return std::move(input);
}

int main() {
    const auto immutable_vec = std::vector<unsigned int>();

    auto mutable_vec = std::vector<unsigned int>();
    mutable_vec = take_and_return(std::move(mutable_vec));

    // Lekker undefined behavior.
    auto foo = Foo(std::move(mutable_vec));
    mutable_vec.push_back(0);
    std::cout << mutable_vec.size() << std::endl;
}

File: (1) borrow_checker_basics/main.rs
#![allow(unused_variables)]


fn add_zero(input: &mut Vec<u32>) {
    input.push(0);
}


fn take_and_return(mut input: Vec<u32>) -> Vec<u32> {
    add_zero(&mut input);
    input
}


fn take_and_keep(input: Vec<u32>) {
    // Do stuff...
}


fn main() {
    let _immutable_vec = Vec::<u32>::new();

    let mut mutable_vec = Vec::<u32>::new();
    mutable_vec = take_and_return(mutable_vec);

    // :)
    // take_and_keep(mutable_vec);
    // mutable_vec.push(0);

    // This is fine.
    take_and_keep(mutable_vec.clone());
    mutable_vec.push(0);
}
